include/vcpkg/binarycaching.h

```diff
@@ -9,53 +9,131 @@
 
 #include <vcpkg/packagespec.h>
 
+#include <iterator>
+#include <string>
 #include <unordered_map>
+#include <vector>
 
 namespace vcpkg
 {
-    struct MergeBinaryProviders;
-
     enum class RestoreResult
     {
-        missing,
-        success,
-        build_failed,
+        unavailable,
+        restored
+    };
+
+    enum class CacheAvailability
+    {
+        unavailable,
+        available
+    };
+
+    enum class CacheStatusState
+    {
+        unknown,   // the cache status of the indicated package ABI is unknown
+        available, // the cache is known to contain the package ABI, but it has not been restored
+        restored   // the cache contains the ABI and it has been restored to the packages tree
+    };
+
+    struct IBinaryProvider;
+
+    struct CacheStatus
+    {
+        CacheStatus() noexcept;
+        CacheStatus(const CacheStatus&);
+        CacheStatus(CacheStatus&&) noexcept;
+        CacheStatus& operator=(const CacheStatus&);
+        CacheStatus& operator=(CacheStatus&&) noexcept;
+        ~CacheStatus();
+
+        bool should_attempt_precheck(const IBinaryProvider* sender) const noexcept;
+        bool should_attempt_restore(const IBinaryProvider* sender) const noexcept;
+
+        bool is_unavailable(size_t total_providers) const noexcept;
+        const IBinaryProvider* get_available_provider() const noexcept;
+        bool is_restored() const noexcept;
+
+        void mark_unavailable(const IBinaryProvider* sender);
+        void mark_available(const IBinaryProvider* sender) noexcept;
+        void mark_restored() noexcept;
+
+    private:
+        CacheStatusState m_status;
+
+        union
+        {
+            // The set of providers who know they do not have the associated cache entry.
+            // Flat vector because N is tiny.
```

Comment:
```suggestion
            // Flat vector set because N is tiny.
```

IMO this makes this line less confusing (this line confused me)